File: app/src/main/java/com/example/david/lists/data/remote/IRemoteRepositoryContract.java
package com.example.david.lists.data.remote;

import com.example.david.lists.data.datamodel.Item;
import com.example.david.lists.data.datamodel.UserList;

import java.util.List;

import io.reactivex.Flowable;

public interface IRemoteRepositoryContract {
    interface Repository {
        Flowable<List<UserList>> getUserLists();

        Flowable<List<Item>> getItems(String userListId);

        void addUserList(UserList userList);

        void addItem(Item item);

        void deleteUserLists(List<UserList> userListList);

        void deleteItems(List<Item> itemList);

        void renameUserList(String userListId, String newName);

        void renameItem(String itemId, String newName);

        void updateUserListPosition(UserList userList, int oldPosition, int newPosition);

        void updateItemPosition(Item item, int oldPosition, int newPosition);

        Flowable<List<UserList>> getEventUserListDeleted();
    }

    interface SnapshotListener {
        Flowable<List<UserList>> getUserListFlowable();

        Flowable<List<Item>> getItemFlowable(String userListId);

        Flowable<List<UserList>> getDeletedUserLists();
    }
}

File: app/src/main/java/com/example/david/lists/data/remote/IRemoteRepositoryContract.kt
package com.example.david.lists.data.remote

import com.example.david.lists.data.datamodel.Item
import com.example.david.lists.data.datamodel.UserList

import io.reactivex.Flowable

interface IRemoteRepositoryContract {
    interface Repository {
        val userLists: Flowable<List<UserList>>

        val userListDeletedObservable: Flowable<List<UserList>>

        fun getItems(userListId: String): Flowable<List<Item>>

        fun addUserList(userList: UserList)

        fun addItem(item: Item)

        fun deleteUserLists(userListList: List<UserList>)

        fun deleteItems(itemList: List<Item>)

        fun renameUserList(userListId: String, newName: String)

        fun renameItem(itemId: String, newName: String)

        fun updateUserListPosition(userList: UserList, oldPosition: Int, newPosition: Int)

        fun updateItemPosition(item: Item, oldPosition: Int, newPosition: Int)
    }

    interface SnapshotListener {
        val userListFlowable: Flowable<List<UserList>>

        val deletedUserListsFlowable: Flowable<List<UserList>>

        fun getItemFlowable(userListId: String): Flowable<List<Item>>
    }
}
